src/runner/runner.go

```diff
@@ -0,0 +1,231 @@
+	RefreshTimeout time.Duration // How often to refresh config
+	MetricsPath    string        // Where to dump metrics to
+}
+
+// Runner executes jobs according to the (fetched from remote) configuration
+type Runner struct {
+	configPaths    []string
+	backupConfig   []byte
+	refreshTimeout time.Duration
+	metricsPath    string
+
+	currentRawConfig []byte // currently applied config
+
+	log *logs.Logger
+
+	stop chan interface{}
+}
+
+// New runner according to the config
+func New(cfg *Config, l *logs.Logger) (*Runner, error) {
+	return &Runner{
+		configPaths:    strings.Split(cfg.ConfigPaths, ","),
+		backupConfig:   cfg.BackupConfig,
+		refreshTimeout: cfg.RefreshTimeout,
+		metricsPath:    cfg.MetricsPath,
+
+		log:  l,
+		stop: make(chan interface{}),
+	}, nil
+}
+
+// Run the runner and block until Stop() is called
+func (r *Runner) Run() {
+	clientID := uuid.New().String()
+	refreshTimer := time.NewTicker(r.refreshTimeout)
+
+	var (
+		stop   bool
+		ctx    context.Context
+		cancel context.CancelFunc
+		wg     sync.WaitGroup
+	)
+
+	for !stop {
+		newRawConfig, err := fetchConfig(r.configPaths, r.log)
+		if err != nil {
+			newRawConfig = r.backupConfig
+		}
+
+		if !bytes.Equal(r.currentRawConfig, newRawConfig) { // Only restart jobs if the new config differs from the current one
+			r.log.Info("New config received, applying")
+
+			var config struct {
+				Jobs []jobs.Config
+			}
+
+			if err := json.Unmarshal(newRawConfig, &config); err != nil {
+				r.log.Warning("Failed to unmarshal job configs: %v", err)
+			} else {
+				if cancel != nil {
+					cancel()
+				}
+
+				ctx, cancel = context.WithCancel(context.Background())
+
+				for i := range config.Jobs {
+					job, ok := jobs.Get(config.Jobs[i].Type)
+					if !ok {
+						r.log.Warning("No such job %q", config.Jobs[i].Type)
+
+						continue
+					}
+
+					if config.Jobs[i].Count < 1 {
+						config.Jobs[i].Count = 1
+					}
+
+					for j := 0; j < config.Jobs[i].Count; j++ {
+						wg.Add(1)
+
+						go func(i int) {
+							job(ctx, r.log, config.Jobs[i].Args)
+							wg.Done()
+						}(i)
+					}
+				}
+
+				r.currentRawConfig = newRawConfig
+
+				r.log.Info("Jobs (re)started")
+			}
+		}
+
+		// Wait for refresh timer or stop signal
+		select {
+		case <-refreshTimer.C:
+		case <-r.stop:
+			refreshTimer.Stop()
+
+			stop = true
+		}
+
+		dumpMetrics(r.log, r.metricsPath, "traffic", clientID)
+	}
+
+	if cancel != nil {
+		cancel()
+	}
+
+	wg.Wait()
+}
+
+// Stop runner asynchronously
+func (r *Runner) Stop() { close(r.stop) }
+
+func fetchConfig(paths []string, l *logs.Logger) ([]byte, error) {
+	for i := range paths {
+		res, err := fetchConfigOnePath(paths[i])
+		if err != nil {
+			l.Warning("Failed to fetch config from %q: %v", paths[i], err)
+			continue
+		}
+
+		l.Info("Loading config from %q", paths[i])
+
+		return res, nil
+	}
+
+	return nil, errors.New("config fetch failed")
+}
+
+func fetchConfigOnePath(path string) ([]byte, error) {
```

Comment:
Not required for merge, but if you get back to this PR before I get to my laptop - would be nice to rename it to something like `fetchSingleConfig`

Reply:
No probs